logger.js

```diff
@@ -85,58 +85,56 @@ function pinoLogger (opts, stream) {
   delete opts.useLevel
 
   const genReqId = reqIdGenFactory(opts.genReqId)
-  loggingMiddleware.logger = logger
-  return loggingMiddleware
-
-  function onResFinished (err) {
-    this.removeListener('close', onResFinished)
-    this.removeListener('error', onResFinished)
-    this.removeListener('finish', onResFinished)
+  const result = (req, res, next) => {
+    return loggingMiddleware(logger, req, res, next)
+  }
+  result.logger = logger
+  return result
 
-    let log = this.log
-    const responseTime = Date.now() - this[startTime]
-    const level = getLogLevelFromCustomLogLevel(customLogLevel, useLevel, this, err)
+  function onResFinished (res, logger, err) {
+    let log = logger
+    const responseTime = Date.now() - res[startTime]
+    const level = getLogLevelFromCustomLogLevel(customLogLevel, useLevel, res, err)
 
     if (level === 'silent') {
       return
     }
 
-    const req = this[reqObject]
-    const res = this
+    const req = res[reqObject]
 
     const customPropBindings = (typeof customProps === 'function') ? customProps(req, res) : customProps
     if (customPropBindings) {
-      log = this.log.child(customPropBindings)
+      log = logger.child(customPropBindings)
     }
 
-    if (err || this.err || this.statusCode >= 500) {
-      const error = err || this.err || new Error('failed with status code ' + this.statusCode)
+    if (err || res.err || res.statusCode >= 500) {
+      const error = err || res.err || new Error('failed with status code ' + res.statusCode)
 
       log[level](
-        onRequestErrorObject(req, this, error, {
-          [resKey]: this,
+        onRequestErrorObject(req, res, error, {
+          [resKey]: res,
           [errKey]: error,
           [responseTimeKey]: responseTime
         }),
-        errorMessage(req, this, error)
+        errorMessage(req, res, error)
       )
 
       return
     }
 
     log[level](
-      onRequestSuccessObject(req, this, {
-        [resKey]: this,
+      onRequestSuccessObject(req, res, {
+        [resKey]: res,
         [responseTimeKey]: responseTime
       }),
-      successMessage(req, this, responseTime)
+      successMessage(req, res, responseTime)
     )
   }
 
-  function loggingMiddleware (req, res, next) {
+  function loggingMiddleware (logger, req, res, next) {
     let shouldLogSuccess = true
 
-    req.id = genReqId(req, res)
+    req.id = req.id || genReqId(req, res)
```

Comment:
when multiple middleware are registered we have to decide how the req.id is to be set. Convention established here is that the first middleware's genReqId function will set the req.id, subsequent middleware's genReqId functions will be ignored, so the req.id will be set only once per request. 

This may warrant additional discussion.